### ISSUE TITLE
Caching support

### ISSUE THREAD

**Issue by marcelofabri** [SwiftGen/SwiftGen#695]
👋

We're using SwiftGen with a custom template for Lottie files (similar to https://github.com/SwiftGen/SwiftGen/issues/627).

However, Lottie files are somewhat large JSONs (ours range from 12 KB to 300 KB).

We've noticed that SwiftGen always takes ~3s to run because of that. If we remove the Lottie generation, it goes down to 0.1s.

I haven't profiled anything, but I'm assuming that's because it's slow to load + parse such complex JSONs (Xcode just hangs when opening them for example 🤷‍♂️).

Given that, I wonder if it'd make sense to build a caching mechanism to SwiftGen. It could be something similar to [what we have on SwiftLint](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Models/LinterCache.swift), based on the last modified date from the template + JSON file.

I've tried using input + output files with `USE_RECURSIVE_SCRIPT_INPUTS_IN_SCRIPT_PHASES` and it correctly skips SwiftGen, but looks like incremental builds take 10-20s longer for us with that enabled.

**Comment by AliSoftware** [comment]
Hello @marcelofabri 

I'm surprised that your execution of `swiftgen` re-parses and re-generates your Lottie files on every run. We're supposed to not re-write the output file if it's unchanged already.

I think what might be happening is that our caching policy so far still re-parses the input files every time and generate the string content for the output every time, and only then does it decide to not write the output file if the string content is the same as the existing file though. We implemented that so that Xcode doesn't risk to `touch` the output file on every run and then risks to abort the builds with "file was modified while being compile" kind of errors back in the day.

Of course, it would be better to have a smarter caching mechanism that enters into action way sooner in the process, especially using the last modified date of template + input file like you suggest indeed to avoid useless parsing altogether 👍

**Comment by marcelofabri** [comment]
Sorry if I wasn’t clear, but it’s not actually rewriting the file (so it’s not triggering a recompilation). It’s just that the processing is slow.

**Comment by marcelofabri** [comment]
Any pointers on how to implement this? I might give it a shot 🚀

**Comment by AliSoftware** [comment]
@marcelofabri Thanks for willing to give it a shot!

So, as an overview of what that would entail:

SwiftGen is composed of multiple parts:
 - SwiftGen target, the command like which parses the arguments and the config file, and invoke the appropriate parsers
 - SwiftGenKit, which contains the core of SwiftGen with its parsers; each parser reads its input file(s) and generate an intermediate representation (that we call a "context") for that parser entry
 - Then Stencil is our template engine which transforms that "context" + each output template to generate the output code.

Note that one entry can have multiple {template+output file} pairs for the same input (for example to parse inputs only once but generate both `.h` and `.m` files from the same context for the `objc` template).

Also, input files are filtered (so if you provide a directory, it will only parse files in that dir matching the filter) and dirs are parsed recursively, which might complicate the computation of the cache key.

Therefore I see two approaches:

1. I think the caching system we want is just a basic one for now, which means that if either the template or the input file has changed its `lastModifiedDate` since last time it was used by a command, then we redo the full parse+context+generate dance from the start.
2. An alternate, optimised version of a caching system could cache the intermediate context for each given input file, but I'm not sure that's worth going that route vs just storing the `lastModifiedDate`? Since keeping the intermediate context in the cache would only be useful if the input file hasn't changed but the template has, which is rare (it's generally the opposite)

---

If you go with route (1), the caching mechanism would have to sit in the `SwiftGen` target rather than `SwiftGenKit`. **Your best entry point I think would be in `ConfigCommands.swift`, in `ConfigEntry::run()`** where you could compare the input + template dates for each entry, and skip it if no changes.

The potential tricky part could be that **one config entry can have multiple inputs at once** (that will be merged into a single context while parsing), and its resulting intermediate context can be used to **generate not just one, but multiple outputs** (with a different template for each). Also, **the list of inputs might change even if the config file doesn't** because if the user provides a **directory**, we scan it recursively and build the list from that.

---

_Braindump:_

So your cache might use the file name as key and last modified date as value, but then to determine if you need to parse a config entry and rebuild the _context_ of a parser entry, you'll need to check if _any_ of the files (inputs or templates) has changed.
 - If even a single input deep into a recursive dir changed, you need to re-parse and regenerate, but if a template used in one of the output has changed you'll also need to reconstruct the intermediate context as we won't probably cache that context's content
 - You'll also need a entry in your cache checking that the config entry's list of input files (after recursion and filtering) and outputs hasn't changed. Otherwise if your config lists one dir as input, and on run 1 you find 5 input files in that dir, but on run 2 you removed one of those input files from your filesystem while not touching the remaining 4, their lastModifiedDate would be the same but we would still need to re-generate the output so that the context and the generated code don't contain the data from that removed file anymore. And you can't use a checksum of the `swiftgen.yml` file for that purpose, because even if the user doesn't change the config file, the cache might need to be invalidated if a file was added or removed in the filesystem

**Comment by mefilt** [comment]
Hey guys
I made tool witch verify exist new files.

https://github.com/mefilt/CheckSwiftGen/tree/master
